MM-47853 - True Up Review Backend

ConorMacpherson opened the pull request:
#### Summary

We need a way to enable customers to do a true-up review without as much hassle. The current process is:
- The customer's admin takes screenshots of the current system statistics, including seats in use, seats purchased, etc.
- The admin sends these screenshots to their CSM
- CSM reviews the details and will inform the customer about any needs for upgrades

These changes introduce a new, less manual process for true-up reviews (**for systems with diagnostics (telemetry) disabled**).
The new process is as follows:
- The customer's admin goes into the `System Console > Site Statistics OR Team Statistics`
- The customer will click a button to send the details to our data warehouse via telemetry (in some cases this is a one-time override as some customers have telemetry disabled)
  - If the customer is air-gapped, a file will be downloaded (via the front-end). The customer's admin will then need to upload this file to the customer web portal.
- The customer's CSM will review the data via `Looker` and determine if the customer needs to upgrade based on their usage.

The following information will be submitted on true-up review:

- ​​Server ID
- Customer Name (if available)
- License Id
- Number of license seats bought
- License plan (e.g. Pro, Enterprise)
- License expiry date
- Number of current activated users
- Server version
- Authentication features
- Number and which Plugin/integration
- Number of webhooks, ~~Playbook, Boards, Call (not sure if possible) per week/month (the goal is to understand if they use the features or not)~~ **(Most of this seems to be available in plugins, but not directly accessible in mattermost-server)**
- ~~User NPS~~

To use the true-up review endpoints, the user needs to:
- Be a self-hosted user (this feature will not work if the user is on a cloud installation)
- Be logged into an admin account.
- Have a license

True Up Review Due dates are as follows:

Quarter | Date True Up Review Available | Due Date | Date True Up Review is no Longer visible
-- | -- | -- | --
Q1 Jan 1 => March 31 | March 16th | April 15th | April 16th
Q2 April 1 => June 30 | June 15th | July 15th | July 16th
Q3 July 1 => Sept 30 | September 15th | October 15th | October 16th
Q4 Oct 1 => Dec 31 | December 16th | January 15th | January 16th


#### Ticket Link

[MM-47853](https://mattermost.atlassian.net/browse/MM-47853)

#### Changes:

- Two new API endpoints:
  - `POST /license/review` - For submitting a true-up review request (extracts data, sends via telemetry, and responds with the data in JSON format for use on the front end).
  - `GET /license/review/status` - For checking the status of the current quarter's true-up review (due date and completion).
- A util function for determining the next due date
- A store for true-up review completion status/due date 
- Tests

#### Related PRs

Front-End Work: https://github.com/mattermost/mattermost-webapp/pull/11844

#### Release Note

```release-note
NONE
```


[MM-47853]: https://mattermost.atlassian.net/browse/MM-47853?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mattermod commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

neallred commented on api4/license.go:
```diff
@@ -296,3 +299,123 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+	}
+
+	reviewProfile := model.TrueUpReviewProfile{
+		ServerId:               c.App.TelemetryId(),
```
I like the well structured representation of this *internally* within the app. It looks like all existing telemetry is sent as a flat map of string to primitves. Its probably a good idea to do the same here to that its easier to consume for whatever human or code will be consuming this data.

We could still keep the structured internal representation but construct a flat map using a technique like this: https://github.com/mattermost/mattermost-server/blob/master/services/telemetry/telemetry.go#L869 .

Otherwise, if you feel strongly about sending out a structured map, I would make sure that the consumer will be easily able to read it.

ConorMacpherson replied:
No strong feelings about a structured map. I think the only reason I ended up doing this was that I built all of this out _around_ the struct.

I definitely didn't think about the end game in terms of where the telemetry data is sent to. A flat map definitely sounds much safer, especially if that's what we're doing everywhere else. I'll make this change.

neallred reviewed (changes requested):
The approach looks good to me overall. I think we'll want to send out a "flat" representation of the data (or at least make sure the consumer can easily handle a structured representation). Found a few typos we'll want to fix, and an opportunity where we might be able to reuse existing analytics code.

ConorMacpherson commented on api4/license.go:
```diff
@@ -296,3 +299,123 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	outgoingWebhookCount, err := c.App.Srv().Store().Webhook().GetOutgoingTotal()
```
Ah, I didn't realize we marked webhooks as deleted. I'll make this change and remove the functionality I added for getting the counts.

ConorMacpherson commented on api4/license.go:
```diff
@@ -296,3 +299,123 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+	}
+
+	// Authentication Features
+	mfaUsed := c.App.Config().ServiceSettings.EnforceMultifactorAuthentication
```
Definitely! Good point :)

ConorMacpherson commented on model/license.go:
```diff
@@ -39,6 +39,15 @@ var (
 )
 
+const (
+	TrueUpReviewTelemetryName          = "true-up-review-sent"
```
How did I not notice this 😆 Thanks for pointing that out :)

ConorMacpherson commented on api4/license.go:
```diff
@@ -296,3 +299,123 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+
+	var subscription *model.Subscription
+	var err error
+	if c.App.Cloud() != nil {
```
Good point, I think I may have been missing some context here on who actually needs to use this feature. It makes sense that it would be self-hosted only. I'll add the NotImplemented error in the case of cloud users.

ConorMacpherson commented:
Going to move this out of review to add a bit more functionality after some new requirements were added.
The next bit of work will essentially just be checking if the user can/needs to run a true-up review:
- Determining the next due date for true-up review (may actually do this in the front-end
- Ensuring the user can only run one true-up review per quarter before each due date.

ConorMacpherson commented:
/update-branch

neallred reviewed (changes requested):
The approach looks good and code quality looks good. We'll want to check the license for whether it is cloud instead of whether it has the cloud interface. I think we want to collect some less information (detailed below), and I think we want to flatten the auth feature list representation from a list to a CSV string.

ConorMacpherson commented on api4/license.go:
```diff
@@ -296,3 +301,174 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !status.Completed {
+		// Send telemetry data.
+		json.Unmarshal(reviewProfileJson, &telemetryProperties)
+		delete(telemetryProperties, "plugins")
```
Done. Did this for the plugin too.

ConorMacpherson commented on utils/license_test.go:
```diff
@@ -91,3 +92,58 @@ func TestGetLicenseFileFromDisk(t *testing.T) {
+		due = GetNextTrueUpReviewDueDate(now)
+		assert.Equal(t, time.December, due.Month())
+	})
+}
```
Added. I didn't think it would do this either, but it does.

neallred replied:
Heh, sure enough: https://pkg.go.dev/time#Date

AGMETEOR commented on api4/license.go:
```diff
@@ -296,3 +304,194 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(err, &nfErr):
+			c.Err = model.NewAppError("requestTrueUpReview", "api.license.true_up_review.status_not_found", nil, "Could not find any true up status records", http.StatusNotFound).Wrap(err)
```
Maybe we should just log (not at error level)

ConorMacpherson commented on api4/license.go:
```diff
@@ -296,3 +304,194 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(err, &nfErr):
+			c.Err = model.NewAppError("requestTrueUpReview", "api.license.true_up_review.status_not_found", nil, "Could not find any true up status records", http.StatusNotFound).Wrap(err)
```
I'm a bit confused about option 2.
Are you saying to retry the creation if the first one failed? Or did you mean if `GetTrueUpReviewStatus` fails, initialize a new record and create it in the DB through `CreateTrueUpReviewStatusRecord`?

ConorMacpherson replied:
I think I've got what you're looking for here covered. If no record can be found, we create a new one with the next due date and the `complete` status set to `false`.

AGMETEOR replied:
Yes, you have covered it. Thanks.

neallred commented on api4/license.go:
```diff
@@ -296,3 +301,174 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+	}
+
+	reviewProfile := model.TrueUpReviewProfile{
+		ServerId:               c.App.TelemetryId(),
```
username and password are things we wouldn't want to expose (and aren't exposed here). Personally identifiable information means sensitive or non-sensitive information that can be used to tie data to people in real life.

At most, I think this is something we would want to ask legal about. It does not seem like a security issue. Wanted to let John reply because the figma shows [some discussion](https://www.figma.com/file/1qWr4RtBjwQZyGQMT7z9Jp?node-id=503:38649#332756441) of consulting with legal already, and I don't want to duplicate conversation or cause confusion if this has already been discussed.
 
My concern is that we shouldn't overcollect information, and that doing so might lead to unfair levels of information or might lead to degraded likelihood of information being kept anonymous. My point is that we don't seem to need some of this information from a functionality perspective, and it seems like we are blurring functionality (is user count compliant with license seat limits) and data collection (various server stats).

One overcollection issue is that some of this information collection (talking about data other than the server id here, like specific plugin usage) isn't directly tied to license seat compliance, and that it could be used by us to infer how much value an organization perceives in its contract, and give Mattermost extra bargaining power that the customer didn't intend to disclose when they just thought they were complying with the license true-up process. At a minimum, that seems like something we would want to call out in out documents on privacy and data collection, if we do indeed want to collect the info.

Disclosing together the server's unique telemetry id and the organization's concrete information makes it so thereafter, any server-side tracking events it fires, we concretely know which organization is emitting them. That's something we don't presently know, and it makes that data that much less anonymous. It could also conceivably be used to gather server-specific usage data to unfairly inform us of the customer's perceived value of using Mattermost.

The other degraded privacy concern is that if the client-side tracking events also contain server telemetry id (to my knowledge, [they do not](https://github.com/mattermost/mattermost-webapp/blob/master/packages/mattermost-redux/src/client/rudder.ts) at the present), or if there is an unanticipated future way to correlate client-side tracking events to a server telemetry id, now whoever knows that information knows concretely which organization any given client-tracking event belongs to. By not collecting server telemetry id in true up, we keep the id more anonymous, which limits how de-anonymized server and client tracking events can ever be.

johndavidlugtu replied:
@neallred 

> are we sure we need/want to collect the server telemetry id?

Preference is to collect it, to keep it consistent with [instances that have telemetry turned on](https://mattermost.looker.com/dashboards/270?Account%20Owner=).

In the linked dashboard ^ , `Instance ID` is server telemetry ID, yes?

Re: the telemetry docs, I'll consult this with the docs team + legal.

On the Privacy Policy though,

> We refer to this information as “telemetry data,” and it does not include any end user personal identifiers or message contents.

This statement holds, as I don't think we collect "end user personal identifiers".

neallred replied:
I don't have access to view that dashboard, but my guess is that the Instance ID displayed there is the telemetry ID.

As follow-up and for transparency, we discussed in a team meeting that:
* We already collect license id together with telemetry id (https://github.com/mattermost/mattermost-server/blob/master/services/telemetry/telemetry.go#L855-L874) when telemetry is on
* If telemetry is off, telemetry events are not sent in the first place, so sending telemetry id and license info together reveal past telemetry data.
* As stated above, we will discuss more with docs/legal team about any changes needed to documentation, privacy policy, and/or data collection proposed in the ticket backing this PR.

johndavidlugtu replied:
> I don't have access to view that dashboard, but my guess is that the Instance ID displayed there is the telemetry ID.

I've asked Daniel Sischy to grant you access

> If telemetry is off, telemetry events are not sent in the first place, so sending telemetry id and license info together reveal past telemetry data.

I'm not sure I understand. If telemetry events aren't sent in the first place, would it not reveal any past telemetry data?
@neallred

ConorMacpherson replied:
I belive telemetry can be enabled and disabled as much as the customer likes, so they may have had it one in the past but now want it disabled @johndavidlugtu

AGMETEOR commented:
Hey @ConorMacpherson code LGTM, but I want to first wait approving because we have a pending discussion initiated by @neallred  regarding the data we want to collect.

johndavidlugtu commented on api4/license.go:
```diff
@@ -296,3 +301,174 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+		delete(telemetryProperties, "plugins")
+		plugins := reviewProfile.Plugins.ToMap()
+		for pluginName, pluginValue := range plugins {
+			telemetryProperties[pluginName] = pluginValue
```
@neallred good callout!

> the plugins and auth features seem like more like telemetry data since as far as I know they aren't needed for checking license compliance

The primary objective here for us is to gather data needed to conduct quarterly true up reviews. Everything else is a bonus intended to help CSMs have more meaningful conversations with customers.

Outside of the two you mentioned (plugins and auth features), are there other things you think are under telemetry data?

neallred replied:
Yeah, the other ones that seem like telemetry more than strictly checking license compliance are:
* server version
* server installation type
* total incoming webhooks
* total outgoing webhooks

As mentioned above, I think even if we decide we want to collect plugin info, we'll want to do similar to how `trackPluginConfig` does it and only collect info on plugins we administer through the Mattermost marketplace.

ConorMacpherson replied:
👍 I'll add a change to ensure we only use plugins administered through the MM Marketplace.

ConorMacpherson replied:
In the case that we need to remove the telemetry data, I've made a PR on another branch with changes to make that happen: https://github.com/mattermost/mattermost-server/pull/21978

Removes:

- server version
- server installation type
- total incoming webhooks
- total outgoing webhooks
- plugin data
- auth features

The true up review "profile" looks like this with the change:

```json
{
	"active_users": 1,
	"customer_name": "Mattermost Dev License",
	"license_id": "m7gqopm2xs30u6q8of7bry4vel",
	"license_plan": "Enterprise Dev",
	"licensed_seats": 200000,
	"server_version": "7.7.0"
}
```

neallred replied:
:100:

johndavidlugtu replied:
@ConorMacpherson @neallred 

So I heard back from legal and we're good to proceed with collecting the following information:

- Server ID
- Customer Name (if available)
- License Id
- Number of license seats bought
- License plan (e.g. Pro, Enterprise)
- License expiry date
- Number of current activated users
- Server version
- Authentication features
- Number and which Plugin/integration
- Number of webhooks

We're covered by existing legalese, specifically the Privacy Policy. Personally Identifiable Information (PII) would pertain to end user information, and not the customer at an account-level.

That said, as we will need to add some documentation on the true-up feature we've developed here, we'll list there the items we are collecting.

Let me know if there's further questions!

cc: @marianunez

neallred commented on services/telemetry/telemetry.go:
```diff
@@ -149,13 +149,13 @@ func (ts *TelemetryService) getRudderConfig() RudderConfig {
 
-func (ts *TelemetryService) telemetryEnabled() bool {
+func (ts *TelemetryService) TelemetryEnabled() bool {
 	return *ts.srv.Config().LogSettings.EnableDiagnostics && ts.srv.IsLeader()
```
I don't think we can use `TelemetryEnabled` **for the API endpoints** because it checks both an application setting **and** status within a cluster of Mattermost servers. In clustered deployments, a client can get inconsistent answers about whether true-up is/should be performed automatically.

So for example, if a customer's deployment has 2 servers to give them high availability, then depending on how load balancing is handled, the user may:
* hit the follower for the get true up status
* be told that telemetry is not enabled
* see and click the true up submit button
* the request is sent to the leader
* the leader fails to update the request since telemetry is enabled.
* The user doesn't see any change of state occur since the endpoint returns that status and doesn't send or complete it.

I think we'll need to check the EnableDiagnostics setting instead. I don't see another helper for checking EnableDiagnostics. 0/5 on whether we want it.

We may not need to return `telemetry_enabled` to the client. It looks like users have access to that config information already:
* definition: https://github.com/mattermost/mattermost-webapp/blob/master/packages/types/src/config.ts#L62
* endpoint /api/v4/config/client?format=old'
* 
It should be accessible through the getConfig selector.

I think we **do** want to use `TelemetryEnabled()` **for the job**, because that way in clustered deployments, we don't get multiple nodes trying to send the true up information.

ConorMacpherson replied:
Make sense. I'll make this change shortly, just running out to get something to eat for lunch.

neallred commented on model/true_up_review_profile.go:
```diff
@@ -0,0 +1,46 @@
+type TrueUpReviewStatus struct {
+	Completed        bool  `json:"complete"`
+	DueDate          int64 `json:"due_date"`
+	TelemetryEnabled bool  `json:"telemetry_enabled"`
```
Because of the other comment about this being available in the client. I don't think we need to add this to the struct here.

neallred reviewed (commented):
Getting close! I think for the API endpoints we need to check the EnableDiagnostics setting but not take cluster leader status into account.

marianunez commented:
> For systems with diagnostics (telemetry) enabled, a daily job has been created, which will run if the following conditions are met:
> 
> * Telemetry is enabled
> * A license is present
> * The license is not a cloud license
> * The license is not a starter/trial license
> For systems with diagnostics (telemetry) enabled, a daily job has been created, which will run if the following conditions are met:
> 
> * Telemetry is enabled
> * A license is present
> * The license is not a cloud license
> * The license is not a starter/trial license
> 
> The job runs once per day and if the above conditions are met, it checks if the current date is within the next true-up review due date (refer to the table below)

@ConorMacpherson if telemetry is enabled, we currently already have a job that sends the server diagnostics information which should be enough for true-up reviews by CSM. https://github.com/mattermost/mattermost-server/blob/ed23df6a2e48a4b229e8fb37c28104304b31b91d/services/telemetry/telemetry.go#L156

Is there anything additional that this additional job would need to do?

ConorMacpherson commented:
@AGMETEOR @marianunez My bad about the job. I was told there was not anything submitting this data. I assumed the same thing too since customers were just sending screenshots before, but obviously there are multiple conditions on that. I'll remove the job portion of this when back online today.

neallred commented:
Follow up on this, I didn't realize the telemetry job already sent license info, so I said we needed the added job (but we don't).

marianunez commented:
> My bad about the job. I was told there was not anything submitting this data. I assumed the same thing too since customers were just sending screenshots before, but obviously there are multiple conditions on that. I'll remove the job portion of this when back online today.

@ConorMacpherson @neallred Apologies for the confusion, I should've mentioned that for telemetry enabled we had this job somewhere in the ticket. I had even envisioned we could reuse some of the logic we had that was sending the daily telemetry but I think you guys have it covered here and it may change more based on what we discussed yesterday.

ConorMacpherson commented:
/update-branch

ConorMacpherson commented:
@neallred @AGMETEOR I had to make a fairly major change to this due to some new clarity on some requirements.

Previously, the due dates were 15 days into the last month of a quarter, and then true up was deemed "expired" after these 15 days, following this table:


Quarter | Date Available | Due Date | Date True Up Review is no Longer visible
-- | -- | -- | --
Q1 | March 1st | March 15th | March 16th
Q2 | June 1st | June 15th | June 16th
Q3 | September 1st | September 15th | September 16th
Q4 | December 1st | December 15th | December 16th

During our feature review today, it was revealed that true-up should be available 15 days before the end of the quarter, as well as 15 days after the end of the quarter.

I've made these changes, ensuring we are following a correct fiscal quarter calendar and adjusted the dates in the tests, with all of them passing. Additionally, I've moved true-up related utils into their own folder. Now, this is the table that should be followed:



Quarter | Date True Up Review Available | Due Date | Date True Up Review is no Longer visible
-- | -- | -- | -- |
Q1 **Jan 1 => March 31** | March 16th | April 15th | April 16th
Q2 **April 1 => June 30** | June 15th | July 15th | July 16th
Q3 **July 1 => Sept 30** | September 15th | October 15th | October 16th
Q4 **Oct 1 => Dec 31** | December 16th | January 15th | January 16th

If you want to re-review these sections of the code, it would be much appreciated :)

ConorMacpherson commented on utils/true_up.go:
```diff
@@ -0,0 +1,84 @@
+	if nowMonth >= time.October && nowMonth <= time.December {
+		finalQuarterYear = nowYear + 1
+	}
+	trueUpSubmissionWindows := []DueDateWindow{
```
2nd Quarter and 3rd Quarter should start on the 15th as their final months have only 30 days in each (June and September)

Your method looks good, but there are a few things I noticed about it. The due date should always be the 15th (15 days after the last end of the quarter), but it's often it's returning the 16th for me.
I think adding two options adds some unnecessary complexity. We really only ever need the due date from the current date and don't exactly need to look try and look for the due date outside of a window or the due date within a window.

We just need to know the next due date from today. If we're past the due date (15th) of the previous quarter, then we find the next due date automatically, since we're already past the window for the last quarter.

ConorMacpherson replied:
Oh, I see. So your method returns the windows start date?

Or is this following `Date True Up Review is no Longer visible` I added on the table above? That column was meant to show when the panel will no longer be visible. I suppose your route makes more sense to me based on that, and the front-end would just check if it's any day before that within the previous 30 days or so.

ConorMacpherson replied:
Gotcha, thank Allan! :)

mattermod commented:
Failed running enterprise tests. @mattermost/core-build-engineers have been notified. Error:  
```404: Project not found```

neallred reviewed (approved):
LGTM, thanks @ConorMacpherson !

marianunez commented on api4/license.go:
```diff
@@ -296,3 +299,94 @@ func getPrevTrialLicense(c *Context, w http.ResponseWriter, r *http.Request) {
+	// Do not send true-up review data if the user has already requested one for the quarter.
+	// And only send a true-up review via as a one-time telemetry request if telemetry is disabled.
+	telemetryEnabled := c.App.Config().LogSettings.EnableDiagnostics
+	if !status.Completed && telemetryEnabled != nil && !*telemetryEnabled {
```
Wondering if we should move the check of `status.Completed` earlier before getting into actually calculating all of the true up data and exist early if it has already completed? Technically it should never happen as I believe the front-end is not showing the panel if status is completed?

ConorMacpherson replied:
Makes sense to do the completed check before most other conditions. I'll move it out of this one.

mattermod commented:
Failed running enterprise tests. @mattermost/core-build-engineers have been notified. Error:  
```404: Project not found```

mattermod commented:
Failed running enterprise tests. @mattermost/core-build-engineers have been notified. Error:  
```404: Project not found```

mattermod commented:
Failed running enterprise tests. @mattermost/core-build-engineers have been notified. Error:  
```404: Project not found```

ConorMacpherson commented:
/update-branch

yasserfaraazkhan commented:
/update-branch

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/289984) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/289984)
